pkg/notes/notes_map.go

```diff
@@ -0,0 +1,168 @@
+package notes
+
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+)
+
+// MapProvider interface that obtains release notes maps from a source
+type MapProvider interface {
+	// GetMaps
+	GetMaps() ([]*ReleaseNotesMap, error)
+	// Get a ReleaseNotesMap for a specific commit
+	GetMap(int) (*ReleaseNotesMap, error)
+}
+
+// NewProviderFromInitString creates a new map provider from an initialization string
+func NewProviderFromInitString(initString string) (MapProvider, error) {
+	// DirectoryProvider
+	if initString[0:4] == "dir" {
+		parts := strings.Split(initString, ":")
+		if len(parts) != 2 {
+			return nil, errors.New("map provider initialization string is not well formed")
+		}
+
+		return &DirectoryMapProvider{
+			Path: parts[1],
+		}, nil
+	}
+
+	return nil, errors.New("Unkown map provider in init string")
+}
+
+// ParseReleaseNotesMap Parses a Release Notes Map
+func ParseReleaseNotesMap(mapPath string) (notemap *ReleaseNotesMap, err error) {
+	notemap = &ReleaseNotesMap{}
+	yamlData, err := ioutil.ReadFile(mapPath)
+	if err != nil {
+		return notemap, errors.Wrap(err, "reading map yaml")
+	}
+	if err := yaml.Unmarshal(yamlData, &notemap); err != nil {
+		return notemap, errors.Wrap(err, "while unmarshaling map yaml")
+	}
+
+	// PR number is always required
+	if notemap.PR == 0 {
+		return nil, errors.New(fmt.Sprintf("Note map at %s does not have a PR number", mapPath))
+	}
+	// logrus.Infof("Note Value:\n%+v", notemap)
+	return notemap, nil
+}
+
+// ReleaseNotesMap Represents
+type ReleaseNotesMap struct {
+	// Pull request where the note was published
+	PR int `json:"pr"`
+	// SHA of the notes commit
+	Commit      string `json:"commit"`
+	ReleaseNote struct {
```

Comment:
Do you think it would be possible to re-use the `ReleaseNotes` struct there https://github.com/kubernetes/release/blob/master/pkg/notes/notes.go#L55.

Reply:
After I sent this last night I was thinking that maybe we should have all fields in the struct be pointers in order be able to have them defined as nil by default and test if the the value has been overridden.

For example: if read the map and get `Areas` is an empty slice (it's empty value), does this mean that the map wants to modify Alias to be an empty slice? Or simply Alias was not defined? With a pointer, I could do `if Alias != nil`  and then know.

Reply:
Yes, to be able to actually diff something I guess pointers are the right choice. :+1:

Reply:
OK, I've changed all the values to pointers to be able to `if nil` them. I had one doubt regarding this comment: you mean moving all values (Text, Author, etc) to the root of the doc and eliminate the `release-note` node? Or just implementing all values in the struct? I did not implement them all because I figured some of them were never going to be overridden (like `AuthorURL` which you can infer from author) .

Reply:
> OK, I've changed all the values to pointers to be able to `if nil` them. I had one doubt regarding this comment: you mean moving all values (Text, Author, etc) to the root of the doc and eliminate the `release-note` node? Or just implementing all values in the struct? I did not implement them all because I figured some of them were never going to be overridden (like `AuthorURL` which you can infer from author) .

I think it's fine like it is for now. No let's do not remove the `release-note` subtype. :+1: